addon/-private/factory-for.js

```diff
@@ -2,13 +2,30 @@
   Utility to use `ApplicationInstance#factoryFor` if available (2.12+) and
   fallback to private `_lookupFactory` otherwise
 */
-export default function factoryFor(owner, type) {
-  let factory;
+class LegacyFactoryForWrapper {
+  constructor(factory) {
+    this._factory = factory;
+  }
+
+  get class() {
+    return this._factory;
+  }
+
+  create(fullName, options) {
+    if (this._factory) {
+      return this._factory.create(fullName, options);
+    }
+  }
+}
+
+export default function factoryFor(owner, fullName) {
   if (owner.factoryFor) {
-    let maybeFactory = owner.factoryFor(type);
-    factory = maybeFactory && maybeFactory.class;
```

Comment:
The bug stemmed from this since we end up trying to instantiate object from `maybeFactory.class` which do not carry forward the owner.  What we want is really to return `maybeFactory`.